File: 20-concurrency.py
# Python asyncio 適合示範 async/await，同時處理多個等待中任務。
import asyncio
import time
from concurrent.futures import ThreadPoolExecutor


async def fetch_user(user_id):
    await asyncio.sleep(0.1)

    return {"id": user_id, "name": f"User {user_id}"}


async def main():
    users = await asyncio.gather(
        fetch_user(1),
        fetch_user(2),
        fetch_user(3),
    )

    print(users)


asyncio.run(main())

print("---")


# ThreadPoolExecutor 適合將 blocking function 放入 worker threads。
def slow_double(number):
    time.sleep(0.1)

    return number * 2


with ThreadPoolExecutor(max_workers=3) as executor:
    results = list(executor.map(slow_double, [1, 2, 3]))

print(results)

File: 20-concurrency.rb
# Ruby Thread 可以將幾個慢操作同時開始，再用 value 等結果。
def fetch_user(user_id)
  sleep 0.1

  { id: user_id, name: "User #{user_id}" }
end

threads = [1, 2, 3].map do |user_id|
  Thread.new do
    fetch_user(user_id)
  end
end

users = threads.map(&:value)

p users

puts '---'

# CPU/IO demo 都一樣：開 worker thread，最後集中收結果。
def slow_double(number)
  sleep 0.1

  number * 2
end

worker_threads = [1, 2, 3].map do |number|
  Thread.new do
    slow_double(number)
  end
end

results = worker_threads.map(&:value)

p results
